### PR TITLE
MINIFICPP-1995 Add configuring path for flowfile_checkpoint directory

### PR DESCRIPTION
**Pull request by lordgamez**
Added `nifi.flowfile.checkpoint.directory.default` property with default value `${MINIFI_HOME}/flowfile_checkpoint`
If repositories are set to empty value the default value is used instead of failure.

https://issues.apache.org/jira/browse/MINIFICPP-1995

-------------------
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: encrypt-config/tests/resources/minifi.properties [hunk 1]
```diff
@@ -24,6 +24,7 @@ nifi.provenance.repository.directory.default=${MINIFI_HOME}/provenance_repositor
 nifi.provenance.repository.max.storage.time=1 MIN
 nifi.provenance.repository.max.storage.size=1 MB
 nifi.flowfile.repository.directory.default=${MINIFI_HOME}/flowfile_repository
+nifi.flowfile.checkpoint.directory.default=${MINIFI_HOME}/flowfile_checkpoint
```
Comment (szaszm) [inline encrypt-config/tests/resources/minifi.properties:27]: I think we can omit this from test resources if it's the same as the default. It makes sense to add to the example config, but in test resources, it's just redundant.

Comment (lordgamez) [reply encrypt-config/tests/resources/minifi.properties]: Updated in 28dfa20abfc159c5da28c85026fae4f459b430bb

File: examples/kubernetes/daemon-set-log-collection/daemon-set-log-collection.yml [hunk 2]
```diff
@@ -30,6 +30,7 @@ data:
     nifi.provenance.repository.max.storage.time=1 MIN
     nifi.provenance.repository.max.storage.size=1 MB
     nifi.flowfile.repository.directory.default=${MINIFI_HOME}/flowfile_repository
+    nifi.flowfile.checkpoint.directory.default=${MINIFI_HOME}/flowfile_checkpoint
```
Comment (szaszm) [inline examples/kubernetes/daemon-set-log-collection/daemon-set-log-collection.yml:33]: I think the property is not relevant here, so I would leave it out from the example. I would ideally also remove all other properties that are using their default values, to keep the example minimal, but that doesn't need to happen in this PR.

File: examples/kubernetes/sidecar-log-collection/sidecar-log-collection.yml [hunk 3]
```diff
@@ -32,6 +32,7 @@ data:
     nifi.provenance.repository.max.storage.time=1 MIN
     nifi.provenance.repository.max.storage.size=1 MB
     nifi.flowfile.repository.directory.default=${MINIFI_HOME}/flowfile_repository
+    nifi.flowfile.checkpoint.directory.default=${MINIFI_HOME}/flowfile_checkpoint
```
Comment (szaszm) [inline examples/kubernetes/sidecar-log-collection/sidecar-log-collection.yml:35]: I think the property is not relevant here, so I would leave it out from the example. I would ideally also remove all other properties that are using their default values, to keep the example minimal, but that doesn't need to happen in this PR.

File: CONFIGURE.md [hunk 4]
```diff
@@ -119,23 +119,23 @@ for TCP and secure HTTPS communications.
             name: tominifi
             max concurrent tasks: 2
             Properties:
-		SSL Context Service: SSLServiceName
```
[the remark this answers is not in the document]
Comment (lordgamez) [reply CONFIGURE.md:122]: Good catch, updated in cbac8731aa8f068600e1cb76a7fc0b4bdc6e72fc

File: examples/kubernetes/daemon-set-log-collection/daemon-set-log-collection.yml (hunk 2 above)
Comment (lordgamez) [reply examples/kubernetes/daemon-set-log-collection/daemon-set-log-collection.yml:33]: Added jira ticket for this impovement: https://issues.apache.org/jira/browse/MINIFICPP-2004

File: examples/kubernetes/sidecar-log-collection/sidecar-log-collection.yml (hunk 3 above)
Comment (lordgamez) [reply examples/kubernetes/sidecar-log-collection/sidecar-log-collection.yml:35]: Added jira ticket for this impovement: https://issues.apache.org/jira/browse/MINIFICPP-2004

**Review by szaszm — APPROVED** [review]

File: extensions/rocksdb-repos/FlowFileRepository.h [hunk 5]
```diff
@@ -100,11 +100,17 @@ class FlowFileRepository : public ThreadedRepository, public SwapManager {
...
+    if (configure->get(Configure::nifi_flowfile_repository_directory_default, value) && !value.empty()) {
       directory_ = value;
     }
     logger_->log_debug("NiFi FlowFile Repository Directory %s", directory_);
 
+    value.clear();
+    if (configure->get(Configure::nifi_flowfile_checkpoint_directory_default, value) && !value.empty()) {
+      checkpoint_dir_ = value;
+    }
+    logger_->log_debug("NiFi FlowFile Checkpoint Directory %s", checkpoint_dir_);
```
Comment (fgerlits) [inline extensions/rocksdb-repos/FlowFileRepository.h:112]: I think you need to either add `.string()` here, or make the conversion templates in `Logger.h` smarter to accept `path`s.

Comment (lordgamez) [reply extensions/rocksdb-repos/FlowFileRepository.h:112]: Added fix in 9b7fd9553203ec6bfae427546e0fa214e695f17f

**Review by fgerlits — APPROVED** [review]

File: libminifi/include/core/logging/Logger.h [hunk 6]
```diff
@@ -72,6 +76,8 @@ inline decltype(auto) conditional_convert(const Arg& val) {
     return val;
   } else if constexpr (std::is_same_v<decltype(std::declval<const Arg&>().c_str()), const char*>) {
     return val.c_str();
+  } else if constexpr (std::is_same_v<decltype(std::declval<const Arg&>().string()), std::string>) {
+    return val.string().c_str();
```
Comment (adamdebreceni) [inline libminifi/include/core/logging/Logger.h:80]: the `::string() const` method could return a temporary which is destroyed before using the result of `c_str()` leaving us with UB, I think for this to be safe, the `conditional_stringify`  functions should call the `::string() const` method, the returned temporary string is kept alive until the end of the full expression the `conditional_stringify` is part of, which is after `format_string` returns

Comment (lordgamez) [reply libminifi/include/core/logging/Logger.h]: Good catch, updated in fb658f6abd33164157570344436dddae13700823

**Review by adamdebreceni — APPROVED** [review]

**Comment by fgerlits** [comment]
Unfortunately, the code did not compile on Windows after 373dbdd98b16d1a45788c5bb10c926368af0c891, so I have reverted both c07ee3e10c54a19820f7b5ac8d2ee63768ed5759 and 373dbdd98b16d1a45788c5bb10c926368af0c891, and added a `.string()` to the log line: 378501d61d0d0a98c1688f28c7a35161af5e0d3a

**Comment by lordgamez** [comment]
> Unfortunately, the code did not compile on Windows after [373dbdd](https://github.com/apache/nifi-minifi-cpp/commit/373dbdd98b16d1a45788c5bb10c926368af0c891), so I have reverted both [c07ee3e](https://github.com/apache/nifi-minifi-cpp/commit/c07ee3e10c54a19820f7b5ac8d2ee63768ed5759) and [373dbdd](https://github.com/apache/nifi-minifi-cpp/commit/373dbdd98b16d1a45788c5bb10c926368af0c891), and added a `.string()` to the log line: [378501d](https://github.com/apache/nifi-minifi-cpp/commit/378501d61d0d0a98c1688f28c7a35161af5e0d3a)

Thanks for the update, you are right it's better to keep it simple at this point.

**Pull request closed without merging**